rats-apps/test/python/rats_test/apps/test_annotated_container.py

```diff
@@ -0,0 +1,288 @@
+from collections.abc import Callable
+from typing import Any, NamedTuple, ParamSpec
+
+import pytest
+
+from rats import apps
+
+
+class ExampleStorage:
+
+    called: int
+    data: list[str]
+
+    def __init__(self) -> None:
+        self.called = 0
+        self.data = []
+
+    def save(self, data: str) -> None:
+        self.called += 1
+        self.data.append(data)
+
+
+class ExampleConfig(NamedTuple):
+    account: str
+    table: str
+    max: int
+
+
+class ExampleExecutable:
+
+    called: int
+
+    def __init__(self) -> None:
+        self.called = 0
+
+    def execute(self) -> None:
+        self.called += 1
+
+
+P_ProviderParams = ParamSpec("P_ProviderParams")
+
+
+def custom(callback: Callable[[P_ProviderParams], apps.ServiceId]) -> Callable[[P_ProviderParams], apps.ServiceId]:
+    def run(*args, **kwargs) -> apps.ServiceId:
+        return callback(*args, **kwargs)
+
+    return run
+
+
+class ExampleGroups:
+    STORAGE_REPLICAS = apps.ServiceId[ExampleStorage]("/group/storage-replicas")
+    EXECUTABLES = apps.ServiceId[ExampleExecutable]("/group/exe")
+    MISSING = apps.ServiceId[ExampleStorage]("/group/missing")
+
+
+class ExampleIds:
+    STORAGE = apps.ServiceId[ExampleStorage]("/service/storage")
+    EXECUTABLE = apps.ServiceId[ExampleExecutable]("/service/exe")
+    MISSING = apps.ServiceId[Any]("/service/missing")
+    DUPLICATE = apps.ServiceId[Any]("/service/duplicate")
+    GROUPS = ExampleGroups
+
+    make_service_id = custom(lambda x: apps.ServiceId[Any](f"/generated/{x}"))
+    make_config_id = custom(lambda x: apps.ConfigId[Any](f"/generated/config-{x}"))
+
+
+class ExampleContainer(apps.AnnotatedContainer):
+
+    @apps.service(ExampleIds.STORAGE)
+    def storage(self) -> ExampleStorage:
+        return ExampleStorage()
+
```

Comment:
we had the notation of private and public services.
perhaps we should consider something like

`
@app.service
def my_service_provider_method(self) -> MyService:
    ...
`
for creating a private service, without need to name it.  the decorator can automatically create a service id from module+class+method names.

Within the module we can refer to this service as `@apps.service_id(MyContainer.my_service_provider_method)`, and in particular we can create a public service id using:

`
class ExampleServiceIds:
    MY_SERVICE = @apps.service_id(MyContainer.my_service_provider_method)
`

Reply:
i tried doing something like that, but ended up with some cyclical dependencies i wasn't able to untangle. can we try to add something like that in a future PR? for now we should be able to continue using the pattern of creating an id mapping class that we never expose in `__init__`